cmd/federation-controller/main.go

```diff
@@ -205,19 +215,41 @@ func main() {
 		}
 	}
 
-	istioClient, err := istiokube.NewClient(istiokube.NewClientConfigForRestConfig(kubeConfig), "")
-	if err != nil {
-		log.Fatalf("failed to create Istio client: %v", err)
-	}
-
-	rm := kube.NewReconcilerManager(
-		meshConfigPushRequests,
+	reconcilers := []kube.Reconciler{
 		kube.NewGatewayResourceReconciler(istioClient, istioConfigFactory),
 		kube.NewServiceEntryReconciler(istioClient, istioConfigFactory),
 		kube.NewWorkloadEntryReconciler(istioClient, istioConfigFactory),
-		kube.NewVirtualServiceReconciler(istioClient, istioConfigFactory),
-		kube.NewDestinationRuleReconciler(istioClient, istioConfigFactory),
-	)
+		kube.NewPeerAuthResourceReconciler(istioClient, namespace),
+	}
+	if cfg.MeshPeers.Local.IngressType == config.OpenShiftRouter {
+		routeClient, err := routev1client.NewForConfig(kubeConfig)
+		if err != nil {
+			log.Fatalf("failed to create Route client: %v", err)
+		}
+
+		reconcilers = append(reconcilers, kube.NewDestinationRuleReconciler(istioClient, istioConfigFactory))
+		reconcilers = append(reconcilers, kube.NewEnvoyFilterReconciler(istioClient, istioConfigFactory))
+		reconcilers = append(reconcilers, kube.NewRouteReconciler(routeClient, openshift.NewConfigFactory(*cfg, serviceLister)))
+
+		go func() {
+			log.Debugf("Resolving %s", cfg.MeshPeers.Remote.Addresses[0])
+			lastIPs := networking.Resolve(cfg.MeshPeers.Remote.Addresses[0])
+			for {
+				log.Debugf("Resolving %s", cfg.MeshPeers.Remote.Addresses[0])
+				ips := networking.Resolve(cfg.MeshPeers.Remote.Addresses[0])
+				sort.Strings(ips)
+				if !slices.Equal(lastIPs, ips) {
+					log.Infof("IP addresses of %s have changed", cfg.MeshPeers.Remote.Addresses[0])
+					lastIPs = ips
+					meshConfigPushRequests <- xds.PushRequest{TypeUrl: xds.ServiceEntryTypeUrl}
+					meshConfigPushRequests <- xds.PushRequest{TypeUrl: xds.WorkloadEntryTypeUrl}
+				}
+				time.Sleep(1 * time.Second)
```

Comment:
I think we will refactor this goroutine when we will have kube-builder and all its components, so I want simplicity for now. I also think that fine grained settings like this one should be exposed not earlier than in tech-preview release, because we may remove it or completely change in the meantime.